pkg/runner/mounting_runner.go

```diff
@@ -0,0 +1,75 @@
+package runner
+
+import (
+	"context"
+	"path/filepath"
+
+	"github.com/buildbarn/bb-remote-execution/pkg/proto/configuration/bb_runner"
+	runner_pb "github.com/buildbarn/bb-remote-execution/pkg/proto/runner"
+	"github.com/buildbarn/bb-storage/pkg/filesystem"
+	"github.com/buildbarn/bb-storage/pkg/filesystem/path"
+	"github.com/buildbarn/bb-storage/pkg/util"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+type mountingRunner struct {
+	base           runner_pb.RunnerServer
+	buildDirectory filesystem.Directory
+	mount          *bb_runner.InputMountOptions
+}
+
+// NewMountingRunner is a decorator for Runner
+// that mounts `mount` before running a build action.
+//
+// This decorator can be used for chroot runners
+// that must mount special filesystems into the input root.
+func NewMountingRunner(base runner_pb.RunnerServer, buildDirectory filesystem.Directory, mount *bb_runner.InputMountOptions) runner_pb.RunnerServer {
+	return &mountingRunner{
+		buildDirectory: buildDirectory,
+		mount:          mount,
+		base:           base,
+	}
+}
+
+func (r *mountingRunner) Run(ctx context.Context, request *runner_pb.RunRequest) (response *runner_pb.RunResponse, err error) {
+	pathResolver := buildDirectoryPathResolver{
+		stack: util.NewNonEmptyStack(filesystem.NopDirectoryCloser(r.buildDirectory)),
+	}
+	scopeWalker := path.NewRelativeScopeWalker(&pathResolver)
+	defer pathResolver.closeAll()
```

Comment:
This might keep more file descriptors open than strictly necessary. Would there be a way for us to close intermediate directories while `Run()` is called?